### PR TITLE
Wsj 2mix transformer

### PR DESCRIPTION
**Pull request by simpleoier**

### CODE REVIEW

File: egs/wsj/asr1/conf/train_multispkr_transformer.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+# minibatch related
+batch-size: 16
+maxlen-in: 512  # if input length  > maxlen-in, batchsize is automatically reduced
+maxlen-out: 150 # if output length > maxlen-out, batchsize is automatically reduced
+
+# optimization related
+sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
+opt: noam
+accum-grad: 2
```
Comment (sw005320) [inline egs/wsj/asr1/conf/train_multispkr_transformer.yaml:27]: I feel that `accum-grad` is a bit too small.
Maybe it is OK, but just a note.

File: egs/wsj_mix/asr1/RESULTS.md [hunk 2]
```diff
@@ -3,21 +3,38 @@
 ## CER
```
Comment (sw005320) [inline egs/wsj_mix/asr1/RESULTS.md:3]: Maybe we can just provide the averaged results?
Also, is it possible to follow the result form used in the other recipes?
Please check https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#how-to-contribute-to-espnet

**Comment by sw005320** [comment]
@sas91, could you also review this PR?

**Comment by sas91** [comment]
> @sas91, could you also review this PR?

Sure I will do it.

File: egs/wsj/asr1/conf/train_multispkr_transformer.yaml (hunk 1 above)
Comment (simpleoier) [reply egs/wsj/asr1/conf/train_multispkr_transformer.yaml:27]: Yes, maybe it is too small. Later, I'll try to set `accum-grad` larger. Anyway, transformer with `accum-grad=2` shows improvement against RNN models.

File: egs/wsj_mix/asr1/RESULTS.md (hunk 2 above)
Comment (simpleoier) [reply egs/wsj_mix/asr1/RESULTS.md]: The new version is updated.

File: espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+from espnet.nets.pytorch_backend.transformer.initializer import initialize
+from espnet.nets.pytorch_backend.transformer.label_smoothing_loss import (
+    LabelSmoothingLoss,  # noqa: H301
+)
+from espnet.nets.pytorch_backend.transformer.mask import subsequent_mask
+from espnet.nets.pytorch_backend.transformer.mask import target_mask
+from espnet.nets.pytorch_backend.transformer.plot import PlotAttentionReport
+
+
+class E2E(ASRInterface, torch.nn.Module):
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py:40]: Can this E2E inherit `e2e_asr_transformer.E2E` to reuse more redundant methods, e.g., `encode`, plots?

File: espnet/nets/pytorch_backend/transformer/encoder.py [hunk 4]
```diff
@@ -187,3 +205,145 @@ def forward_one_step(self, xs, masks, cache=None):
         if self.normalize_before:
             xs = self.after_norm(xs)
         return xs, masks, new_cache
+
+
+class EncoderMix(Encoder, torch.nn.Module):
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/transformer/encoder.py:210]: We should place this class in a new module, for example `encoder_mix.py`.

**Comment by ShigekiKarita** [comment]
Can you setup the integration test in `mini_an4` with a minimum transformer config as well as RNN?
https://github.com/espnet/espnet/blob/d49f10f62144a59d496d0eb796a5ce6a405dcaa1/ci/test_integration.sh#L34-L37

File: espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py [hunk 5]
```diff
@@ -0,0 +1,530 @@
+# Copyright 2019 Shigeki Karita
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+"""Transformer speech recognition model (pytorch)."""
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py:4]: If you can mention your ICASSP2020 paper (https://arxiv.org/pdf/2002.03921.pdf) in this docstring, RESULTS.md (egs), and anywhere, that would be super cool.

**Comment by ShigekiKarita** [comment]
This must be the biggest feature in v.0.8.0! LGTM overall except for the missing integration test.

File: espnet/nets/pytorch_backend/transformer/encoder.py (hunk 4 above)
Comment (simpleoier) [reply espnet/nets/pytorch_backend/transformer/encoder.py]: Done. Thanks!

File: espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py (hunk 3 above)
Comment (simpleoier) [reply espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py]: Good suggestion, thanks!

File: espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py (hunk 5 above)
Comment (simpleoier) [reply espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py:4]: Done. Thanks!

**Comment by sas91** [comment]
@simpleoier "egs/wsj_mix/asr1/run.sh" seems to be missing model averaging in the decode stage while the RESULTS seem to have it.

**Comment by ShigekiKarita** [comment]
@kamo-naoyuki fixed the CI errors related to prebuilt kaldi bin in https://github.com/espnet/espnet/pull/1922 Can you merge the upstream develop?

**Comment by sw005320** [comment]
> @kamo-naoyuki fixed the CI errors related to prebuilt kaldi bin in #1922 Can you merge the upstream develop?

@simpleoier, can you do it?
We are close to finish this PR.

**Comment by simpleoier** [comment]
> @simpleoier, could you restart it?
> Could you also clean up the commit history to only show your update for a code review?

Yeah, sure. I will finish this soon.

File: egs/wsj/asr1/conf/train_multispkr_transformer.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# network architecture
```
Comment (sw005320) [inline egs/wsj/asr1/conf/train_multispkr_transformer.yaml:1]: Is this place OK?
I think it should be under `wsj_mix`

File: egs/wsj_mix/asr1/RESULTS.md [hunk 7]
```diff
@@ -1,23 +1,75 @@
...
+ ```
+ exp/tr_pytorch_train_multispkr_transformer_no_preprocess_delta/decode_cv_decode_lm_word65000_model.last10.avg.best/min_perm_result.json
+ |  # Snt |   Corr     Sub     Del   Ins    Err   |
+ |  503   |   13996    1989    441   317    16.72 |
+ exp/tr_pytorch_train_multispkr_transformer_no_preprocess_delta/decode_tt_decode_lm_word65000_model.last10.avg.best/min_perm_result.wrd.json
+ |  # Snt |   Corr     Sub     Del   Ins    Err   |
+ |  333   |   10094    1055    169   200    12.58 |
+ ```
+
+    - WSJ0\_2mix (model link: https://drive.google.com/open?id=1xm2W1AXBgnccq-AFkp5-sDe2R0X8CMNx)
```
Comment (sw005320) [inline egs/wsj_mix/asr1/RESULTS.md:32]: I don't know why but this line becomes an inline code. Please fix it. 
(see https://github.com/espnet/espnet/blob/d24f502e942ebe0bea7e284a60fb01b8370a8d48/egs/wsj_mix/asr1/RESULTS.md)

File: espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+        :param ndarray x: source acoustic feature (T, D)
+        :return: encoder outputs
+        :rtype: torch.Tensor
+        """
+        self.eval()
+        x = torch.as_tensor(x).unsqueeze(0)
+        enc_output, _ = self.encoder(x, None)
+        return enc_output
+
+    def recog(self, enc_output, recog_args, char_list=None, rnnlm=None, use_jit=False):
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/e2e_asr_mix_transformer.py:228]: One future work (not now) is to implement this beam search under api_v2. 
https://github.com/espnet/espnet/issues/1384
Please check it.

**Comment by JiayiFu** [comment]
I have a naive question: what does the 'mix' in the name mean?

**Comment by simpleoier** [comment]
@JiayiFu It means the multi-speaker speech mixture. For more details, please refer to [this paper](https://arxiv.org/pdf/1811.02062.pdf) for more details.

**Comment by JiayiFu** [comment]
@simpleoier Got it. Thx!

**Pull request closed without merging**